conn_js.go

```diff
@@ -0,0 +1,196 @@
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errIsClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	mut            sync.Mutex
+	incomingData   chan []byte
+	currData       *bytes.Buffer
+	closeSignal    chan struct{}
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
+	}
+
+	if c.currData == nil {
+		select {
+		case data := <-c.incomingData:
+			c.currData = bytes.NewBuffer(data)
+		case <-c.closeSignal:
+			return 0, io.EOF
+		}
+	}
+
+	n, err := c.currData.Read(b)
+	if err == io.EOF {
+		c.currData = nil
+		return n, nil
+	}
+	return n, err
+}
+
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errIsClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	// c.mut.Lock()
```

Comment:
Another reason I called this PR "WIP" is because I need to go through and make sure we are acquiring and releasing a lock when needed. Will take a little bit of careful thought to avoid both race conditions and deadlocks.

Reply:
I addressed this by adding two separate locks. One controls the callbacks for event handlers. We need to call `Release` on these whenever the connection is closed. The other is specifically for controlling access to the buffer which represents incoming data. Let me know what you think.